### PR TITLE
introduce session unstability timeout

### PR DESCRIPTION
**Pull request by himanshug**
allows user to set a "unstableTimeout" (disabled by default)

if set, then once session reaches UNSTABLE state.... it will switch to UNSTABLE_PLUS after unstableTimeout , once UNSTABLE_PLUS is reached then DefaultCopycatClient is closed automatically.

@kuujo it is missing docs/unit-tests which I will add if the approach looks ok to you. and, feel free to suggest alternate names for the things introduced.

### CODE REVIEW

**Comment by kuujo** [comment]
I think this is a fine approach. I don't love the name `UNSTABLE_PLUS`, but I haven't thought of anything better after a few minutes of thesaurusing :-P

I actually wonder if we should just reuse `EXPIRED` for this. What we're effectively doing is  behaving as if the client's session expired after some duration. In fact, we could conceivably use the negotiated `sessionTimeout` to determine when to proactively expire the client's session, or at least default to the session timeout.

**Comment by himanshug** [comment]
yeah, `UNSTABLE_PLUS` is not great but I couldn't think of a better name :)

I did not want to use `EXPIRED` because, that semantically means that server side explicitly expired the session not the client side, so for `EXPIRED` state, `active` is set to `false`. Keeping `EXPIRED` separate keeps semantics clear.
Also, if we reuse `EXPIRED` and user specified RecoveryStrategies.RECOVER .. then there is not much point as, after prolonged UNSTABLE, we will close the session and get back to retrying again.... OTOH I have a wrapper, which on CopycatClient reaching CLOSED state, recreates the client with new set of server addresses and uses that.

**Comment by jhall11** [comment]
What about `STALE` as a name for the new state?

**Comment by kuujo** [comment]
Better than our ideas :-)

**Comment by himanshug** [comment]
@jhall11 wonder why we couldn't think of `STALE` :)

cool, will update the pr with docs and tests now.

**Comment by kuujo** [comment]
+1

File: client/src/main/java/io/atomix/copycat/client/session/ClientSessionState.java [hunk 1]
```diff
@@ -101,10 +109,28 @@ public long getSessionId() {
...
+        unstableSince = System.currentTimeMillis();
+        return setStateAndCallListeners(state);
+      }
     }
+
+    return this;
+  }
+
+  private ClientSessionState setStateAndCallListeners(Session.State state)
+  {
```
Comment (kuujo) [inline client/src/main/java/io/atomix/copycat/client/session/ClientSessionState.java]: Just need to move this bracket up :-)

**Comment by kuujo** [comment]
I think this is great!

File: client/src/main/java/io/atomix/copycat/client/session/ClientSessionState.java (hunk 1 above)
Comment (himanshug) [inline client/src/main/java/io/atomix/copycat/client/session/ClientSessionState.java]: done,

thatz my current intellij formatter , different projects follow different styles :)
I'll need to setup separate formatter for copycat ..... do you guys use intellij ? if yes, can you guys export the formatting spec and put it somewhere?

File: client/src/main/java/io/atomix/copycat/client/DefaultCopycatClient.java [hunk 2]
```diff
@@ -150,6 +154,10 @@ private void onStateChange(Session.State state) {
       case UNSTABLE:
         setState(State.SUSPENDED);
         break;
+      case STALE:
+        setState(State.SUSPENDED);
+        this.close();
```
Comment (jhall11) [inline client/src/main/java/io/atomix/copycat/client/DefaultCopycatClient.java]: Should we try to recover the session here?

Comment (kuujo) [inline client/src/main/java/io/atomix/copycat/client/DefaultCopycatClient.java]: Yeah, it seems to me like this state should be invoking the recovery strategy

Comment (himanshug) [inline client/src/main/java/io/atomix/copycat/client/DefaultCopycatClient.java]: not really, if user wanted recovery to continue happening why would they want to configure the unstabilityTimeout .
If they don't configure unstabilityTimeout then session would never reach STALE state and continue retrying to recover from UNSTABLE state.

**Comment by kuujo** [comment]
Running tests one more time

**Pull request merged**